twitter_api/main.py

```diff
@@ -17,6 +21,164 @@ def before_request():
+    else:
+        abort(401)
+        
+        
+    
+@app.route('/logout', methods=['POST'])
+@has_json_keys('access_token')
+def logout():
+    token = request.json['access_token']
+    g.db.execute('DELETE FROM "auth" WHERE access_token = ?', (token,))
+    g.db.commit()
+    return ('', 204)
+
+
+@app.route('/profile/<username>', methods = ['GET'])
+def get_profile(username):
+    cursor = g.db.execute('SELECT id, username, first_name, last_name, birth_date FROM user WHERE username = ?',(username,))
+    
+    vals = cursor.fetchone()
+    
+    if not vals:
+        abort(404)
+        
+    cols = ["user_id", "username", "first_name", "last_name", "birth_date"]
+    data = dict(zip(cols, vals))
+    cursor = g.db.execute('SELECT  id, content, created FROM tweet WHERE user_id = ? ORDER BY id asc',(data['user_id'],))
+    tweets = [dict(id=row[0], text=row[1], date=row[2], uri="/tweet/" + str(row[0])) for row in cursor.fetchall()]
+    data["tweet"] = tweets
+    data["tweet_count"] = len(tweets)
+    
+    return jsonify(data)
+    
+
+
+@app.route('/profile', methods = ['POST'])
+@json_only
+def update_profile():
+    fields = set(['first_name', 'last_name', 'birth_date'])
+    
+    if fields.intersection(request.json.keys()) != fields:
+        abort(400)
+        
+    if "access_token" not in request.json.keys():
+        abort(401)
+   
+    
+    fn = request.json['first_name']
+    ln = request.json['last_name']
+    bd = request.json['birth_date']
+    token = request.json['access_token']
+    
+    user_id = _find_user_id_by_token(token)
+    
+    g.db.execute("UPDATE user SET first_name=?, last_name=?, birth_date=? WHERE id=?", (fn, ln, bd, user_id))
+    g.db.commit()
+    
+    
+    return('', 201)
+
+def _find_user_id_by_token(token):
+    cursor = g.db.execute("SELECT user_id FROM auth WHERE access_token=?", (token,))
+    return_val = cursor.fetchone()
+    if not return_val:
+        abort(401)
+    return return_val[0]
+    
+def _get_tweet_user_id(tweet_id):
+    cursor = g.db.execute("SELECT user_id FROM tweet WHERE id = ?", (tweet_id,))
+    return_val = cursor.fetchone()
+    
+    if not return_val:
+        abort(404)
+        
+    return return_val[0]
+    
+
+
+    
+    
+@app.route('/tweet/<tweet_id>', methods = ['GET'])
+def get_tweet(tweet_id):
+    cursor = g.db.execute("""SELECT t.id,  t.content, t.created, "/profile/" 
+                                || u.username, "/tweet/" || t.id 
+                                FROM tweet t LEFT JOIN user u ON u.id = t.user_id 
+                                WHERE t.id=?""", (tweet_id,))
+    vals = cursor.fetchone()
+    
+    if not vals:
+        abort(404)
+ 
+    cols = ["id", "content", "date", "profile", "uri"]
+    data = dict(zip(cols, vals))
+    
+    return jsonify(data)
+    
+    
+@app.route('/tweet', methods = ['POST'])
+@json_only
+@has_json_keys('content', error = 400)
+@has_json_keys('access_token', error = 401)
```

Comment:
We need to return different http status codes, depending on which key is missing.